botocore/credentials.py

```diff
@@ -255,27 +270,99 @@ def _seconds_remaining(self):
+        """Check if a refresh is needed.
+
+        A refresh is needed if the expiry time associated
+        with the temporary credentials is less than the
+        provided ``refresh_in``.  If ``time_delta`` is not
+        provided, ``self.advisory_refresh_needed`` will be used.
+
+        For example, if your temporary credentials expire
+        in 10 minutes and the provided ``refresh_in`` is
+        ``15 * 60``, then this function will return ``True``.
+
+        :type refresh_in: int
+        :param refresh_in: The number of seconds before the
+            credentials expire in which refresh attempts should
+            be made.
+
+        :return: True if refresh neeeded, False otherwise.
+
+        """
         if self._expiry_time is None:
             # No expiration, so assume we don't need to refresh.
             return False
 
+        if refresh_in is None:
+            refresh_in = self._advisory_refresh_timeout
         # The credentials should be refreshed if they're going to expire
         # in less than 5 minutes.
-        if self._seconds_remaining() >= self.refresh_timeout:
+        if self._seconds_remaining() >= refresh_in:
             # There's enough time left. Don't refresh.
             return False
-
-        # Assume the worst & refresh.
         logger.debug("Credentials need to be refreshed.")
         return True
 
+    def _is_expired(self):
+        # Checks if the current credentials are expired.
+        return self.refresh_needed(refresh_in=0)
+
     def _refresh(self):
-        if not self.refresh_needed():
+        # In the common case where we don't need a refresh, we
+        # can immediately exit and not require acquiring the
+        # refresh lock.
+        if not self.refresh_needed(self._advisory_refresh_timeout):
             return
 
-        metadata = self._refresh_using()
+        # acquire() doesn't accept kwargs, but False is indicating
+        # that we should not block if we can't acquire the lock.
+        # If we aren't able to acquire the lock, we'll trigger
+        # the else clause.
+        if self._refresh_lock.acquire(False):
+            try:
+                if not self.refresh_needed(self._advisory_refresh_timeout):
+                    return
+                mandatory_refresh = self.refresh_needed(
+                    self._mandatory_refresh_timeout)
+                self._protected_refresh(mandatory=mandatory_refresh)
+                return
+            finally:
+                self._refresh_lock.release()
+        elif self.refresh_needed(self._mandatory_refresh_timeout):
+            # If we're within the mandatory refresh window,
+            # we must block until we get refreshed credentials.
+            with self._refresh_lock:
+                if not self.refresh_needed(self._mandatory_refresh_timeout):
+                    return
+                self._protected_refresh(mandatory=True)
+
+    def _protected_refresh(self, mandatory):
```

Comment:
Ah so ``mandatory`` is suggesting if refresh is mandatory. Maybe call it ``is_mandatory``?